src/engagement_db_to_analysis/engagement_db_to_analysis.py

```diff
@@ -0,0 +1,92 @@
+from os import path
+import json
+from dateutil.parser import isoparse
+
+
+from core_data_modules.logging import Logger
+from core_data_modules.util import TimeUtils
+
+log = Logger(__name__)
+
+
+def get_project_messages_from_engagement_db(pipeline_config, engagement_db, engagement_db_datasets_export_dir):
+    """
+
+    Downloads project messages from engagement database. It performs a full download if there is no previous export and
+    incrementally otherwise.
+
+    :param pipeline_config: Pipeline configuration module.
+    :type pipeline_config: PIPELINE_CONFIGURATION
+    :param engagement_db: Engagement database to update the message in.
+    :type engagement_db: engagement_database.EngagementDatabase
+    :param engagement_db_datasets_export_dir: Directory containing engagement_db dataset files generated by a previous
+                                            run of this pipeline new or changed messages will be updated to these files.
+    :type engagement_db_datasets_export_dir: str
+    :return: engagement_db_dataset_messages_map of engagement_db_dataset to list of messages.
+    :rtype: dict
+    """
+
+    engagement_db_dataset_messages_map = {} # of engagement_db_dataset to list of messages
+
+    for coda_dataset_config in pipeline_config.coda_sync.sync_config.dataset_configurations:
+
+        engagement_db_dataset = coda_dataset_config.engagement_db_dataset
+        previous_export_file_path = path.join(engagement_db_datasets_export_dir, f"{engagement_db_dataset}.json")
+
+        last_updated = None
+        messages = []
+        new_messages = []
+        if path.exists(previous_export_file_path):
+            with open(previous_export_file_path) as f:
+                previous_export = json.load(f)
+
+            for msg in previous_export:
+                if last_updated is None or isoparse(msg["last_updated"]) > last_updated:
+                    last_updated = isoparse(msg["last_updated"])
+
+            incremental_get_filter = lambda q: q \
+                .where("dataset", "==", engagement_db_dataset) \
```

Comment:
This doesn't work if we previously downloaded a message for this dataset but whose dataset has since changed (e.g. because of WS correction). In that case, we need to know that message has been updated and should no longer be included in the result set here, but since it no longer matches this query we won't see it.

We should discuss next week how to make this work; some options I can think of:
- Don't filter by dataset in the cache, which works if the pipeline knows about all active datasets.
- Download all messages in incremental mode, filtering only on timestamp.
- Add a projects list to the messages in the database, then filter where project array-contains this project
- Add a previous_datasets property to messages in the database, then incrementally query for where dataset == db_dataset or previous_datasets array-contains db_dataset

Reply:
Thanks for catching this.  Resolved it splitting the incremental filter into two;  
new_messages_filter - downloads all messages `.where("dataset", "==", engagement_db_dataset).where("message_id", "not-in", cache_message_ids)` and updated_messages_filter  which downloads cache messages that have been updated since the previous download. It then checks for messages whose engagement_db_dataset has changed and logs then and updates those whose other message properties have been updated.

Reply:
I don't think this works at scale, unfortunately, because not-in queries can only check for 10 items at once https://firebase.google.com/docs/firestore/query-data/queries#not-in

Reply:
Thats unfortunate. Updated engagement_db to include previous_dataset array https://github.com/AfricasVoices/Pipeline-Infrastructure/pull/55/files . This now check for `ws_corrected_messages_filter = lambda q: q \
                .where("previous_datasets", "array_contains", engagement_db_dataset) \
                .where("last_updated", ">", latest_message_timestamp)`
to evaluate cache messages to retain in incremental download.